File: frontend/travel-hub/app/lib/types/occupancy.ts
/** Booking occupancy API — snake_case matches backend JSON. */

export type OccupancyLevel = 'HIGH' | 'MEDIUM' | 'LOW';

export interface OccupancyDay {
  date: string;
  total_rooms: number;
  occupied_rooms: number;
  blocked_rooms: number;
  available_rooms: number;
  occupancy_rate: number;
  occupancy_level: OccupancyLevel;
}

export interface OccupancyCalendarResponse {
  property_id: string;
  date_from: string;
  date_to: string;
  days: OccupancyDay[];
}

export interface RoomTypeOccupancyBreakdownRow {
  room_type_id: string;
  room_type_name: string;
  total_rooms: number;
  occupied_rooms: number;
  blocked_rooms: number;
  available_rooms: number;
  occupancy_rate: number;
  occupancy_level: OccupancyLevel;
}

export interface OccupancyDailyBreakdownResponse {
  property_id: string;
  date: string;
  room_types: RoomTypeOccupancyBreakdownRow[];
}

export interface ProjectionDayAlert {
  type: string;
  message: string;
}

export interface OccupancyProjectionDay {
  date: string;
  occupancy_rate: number;
  occupancy_level: OccupancyLevel;
  alert: ProjectionDayAlert | null;
}

export interface LowOccupancyPeriodAlert {
  type: string;
  date_from: string;
  date_to: string;
  average_occupancy_rate: number;
}

export interface OccupancyProjectionResponse {
  property_id: string;
  date_from: string;
  date_to: string;
  days: OccupancyProjectionDay[];
  alerts: LowOccupancyPeriodAlert[];
}

File: backend/services/booking/app/schemas/occupancy.py
from pydantic import BaseModel, ConfigDict, Field


class OccupancyCalendarDayOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    date: str
    total_rooms: int = Field(ge=0)
    occupied_rooms: int = Field(ge=0)
    blocked_rooms: int = Field(ge=0)
    available_rooms: int = Field(ge=0)
    occupancy_rate: float = Field(ge=0.0)
    occupancy_level: str


class OccupancyCalendarResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    property_id: str
    date_from: str
    date_to: str
    days: list[OccupancyCalendarDayOut]


class RoomTypeOccupancyBreakdownOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    room_type_id: str
    room_type_name: str
    total_rooms: int = Field(ge=0)
    occupied_rooms: int = Field(ge=0)
    blocked_rooms: int = Field(ge=0)
    available_rooms: int = Field(ge=0)
    occupancy_rate: float = Field(ge=0.0)
    occupancy_level: str


class OccupancyDailyBreakdownResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    property_id: str
    date: str
    room_types: list[RoomTypeOccupancyBreakdownOut]


class ProjectionDayAlertOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True, protected_namespaces=())

    type: str
    message: str


class OccupancyProjectionDayOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    date: str
    occupancy_rate: float = Field(ge=0.0)
    occupancy_level: str
    alert: ProjectionDayAlertOut | None = None


class LowOccupancyPeriodAlertOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True, protected_namespaces=())

    type: str
    date_from: str
    date_to: str
    average_occupancy_rate: float


class OccupancyProjectionResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    property_id: str
    date_from: str
    date_to: str
    days: list[OccupancyProjectionDayOut]
    alerts: list[LowOccupancyPeriodAlertOut]
